mkosi/distributions/gentoo.py

```diff
@@ -329,50 +321,19 @@ def merge_user_pkgs(self) -> None:
         if self.state.config.packages:
             self.invoke_emerge(pkgs=self.state.config.packages)
 
-
     def invoke_emerge(
         self,
-        inside_stage3: bool = True,
         pkgs: Sequence[str] = (),
         actions: Sequence[str] = (),
         opts: Sequence[str] = (),
     ) -> None:
-        if not inside_stage3:
-            from _emerge.main import emerge_main  # type: ignore
-
-            PREFIX_OPTS: list[str] = []
-            if "--sync" not in actions:
-                PREFIX_OPTS = [
-                    f"--config-root={self.root.resolve()}",
-                    f"--root={self.root.resolve()}",
-                    f"--sysroot={self.root.resolve()}",
-                ]
-
-            MkosiPrinter.print_step(f"Invoking emerge(1) pkgs={pkgs} "
-                                    f"actions={actions} outside stage3")
-            emerge_main([*pkgs, *opts, *actions] + PREFIX_OPTS + self.emerge_default_opts)
-        else:
-            cmd = ["/usr/bin/emerge", *pkgs, *self.emerge_default_opts, *opts, *actions]
+        cmd = ["/usr/bin/emerge", *pkgs, *self.emerge_default_opts, *opts,
```

Comment:
For all other distros, we run the package manager on the host outside of the image, not inside it. I would very much want us to do the same for gentoo. If I read the emerge man page, it seems to support the necessary options (--root and such) so can you modify the gentoo implementation to use emerge installed on the host instead of invoking it inside the image?

Reply:
that would be an almost complete re-write. lets have this merged for now and i will start working on what you're suggesting.

Reply:
> what you're suggesting.
although i'm not sure why that's a better way of doing it. i was actually going to remove `portage` requirement on the host and have the whole thing self-contained.

Reply:
We want to be able to build images without the tools that were used to build them included. Especially for package managers with heavy dependencies, such as emerge and dnf, which pull in python, we want to be able to use those to build images that don't include dnf or emerge so we can keep those images small. While we could try to uninstall the package managers after building the images, it'll be a constant battle to ensure we remove all files associated with the package manager. The other solution is to never install the package manager in the image in the first place, which we can do by running it on the host instead of inside the image.

Reply:
regarding the image size, i am not happy with the result either, and there are lots that can be done there. ATM many, but not all, of those options can be exercised using `mkosi.skeleton/etc/portage/*`. we could expose those but that would possibly require distro-specific flags. i wonder what you guys think?

as for not including the package manager, not an option ATM i'm afraid. upstream ships `portage` unconditionally with their stage3 and i guess sticking with [POLA](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) we want to adhere to that exercise, i would imagine. any thoughts?

there is also an argument to be made from user's point of view. does the user want to have the package manager at its disposal or not. there are many use cases, including my owns, where the answer is Yes. i can think of few cases where that answer is No... it's trade off, as always, so i guess it's better to allow the use to choose.

removing `portage` incidentally doesn't buy us much, low hanging fruits are somewhere else. build dependencies is one, which can be easily done with tweaking the `portage`, tool-chain is another one but do we really want that?
again all of that, IMHO, should be seen from user's point of view; i can't think of many cases where a developer (the main target-user of `mkosi`) wants a final image on the first run, for me at least it takes couple of try to arrive at what i would consider a *golden-master* and even then, at least with Gentoo, things move quickly enough that sometime i even need to start from scratch.

Reply:
The thing is, we want to make mkosi a proper tool for building "golden" non-development images as well, as this is one of the main target use cases aside from the development use case. And to build those images, we should build minimal images by default, without build dependencies, without toolchains, and without heavyweight package managers. Users that want those to be available can then simply add them to the list of packages to be installed.

Reply:
thinking more about this more i have some thoughts to share.

i don't see a way to use host system's `portage` for building images at all and here is why:

allowing users to install their desired packages is an essential part of `mkosi`, i understand. also upstream `toolchain`, the ones shipped within `stage3` root-tree, is a hard dependency for building a consistent os image in case of Gentoo, being a source-based distro.
 
as for image size and what we can shed off:
inclusion of `toolchains` is an upstream-thing, i'm not aware of a smaller system root-tree upstream than `stage3` in fact. note that there is no desktop related stuff (kde/gnome what have you) included in `stage3` so it is, to the best of my knowledge, _the_ smallest we can get from upstream.
yes, there are `stage1` and `stage2` but those are beyond the scope of `mkosi`.

so considering those getting into the business of removing the `toolchains` wouldn't make sense IMHO.
take what we did before `usrmerge` as an example. we used to go to great length to modify the upstream root-tree to make sure `usronly` et al work. that added lots of logic that could be avoided if upstream already provided a root-tree contained under `/usr` which they do now and that reduces code and complexity at our end.

if, one day, upstream provides a root-tree without `toolchains`, considering that they do have a semi-official binary repo now, we can look into using that root-tree.

Reply:
... also note that users can ask portage to not to installed build dependencies of their packages by using `mkosi.skeleton/etc/portage` and/or `mkosi.extra/etc/poratge`. we could document this i guess.

Reply:
I get your point about the toolchains. How is that related to portage though? Can't we build images using portage from the host while still using the toolchains from the image? I understand that probably wouldn't meaningfully impact the image size, but I would still like it that way for consistency with the other distributions.

Reply:
sounds like [cross-compiling](https://wiki.gentoo.org/wiki/Cross_build_environment), possible i guess but complicated.